pkg/runtime/grpc/info.go

```diff
@@ -18,41 +18,57 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
-	"net"
 	"time"
 
-	"google.golang.org/grpc"
-	"google.golang.org/grpc/credentials/insecure"
+	log "github.com/sirupsen/logrus"
 
 	"github.com/inspektor-gadget/inspektor-gadget/internal/deployinfo"
 	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
 )
 
-func loadRemoteDeployInfo() (*deployinfo.DeployInfo, error) {
-	ctx, cancelDial := context.WithTimeout(context.Background(), time.Second*ConnectTimeout)
-	defer cancelDial()
+// InitDeployInfo loads the locally stored deploy info. If no deploy info is stored locally,
+// it will try to fetch it from one of the remotes and store it locally. It will issue warnings on
+// failures.
+func (r *Runtime) InitDeployInfo() {
+	// Initialize info
+	info, err := deployinfo.Load()
+	if err == nil {
+		r.info = info
+		return
+	}
 
-	// Get a random gadget pod and get the info from there
-	pods, err := getGadgetPods(ctx, []string{})
+	info, err = r.loadRemoteDeployInfo()
 	if err != nil {
-		return nil, fmt.Errorf("get gadget pods: %w", err)
+		log.Warnf("could not load gadget info from remote: %v", err)
+		return
 	}
-	if len(pods) == 0 {
-		return nil, fmt.Errorf("no valid pods found to get info from")
+	r.info = info
+
+	err = deployinfo.Store(info)
+	if err != nil {
+		log.Warnf("could not store gadget info: %v", err)
 	}
+}
 
-	pod := pods[0]
-	dialOpt := grpc.WithContextDialer(func(ctx context.Context, s string) (net.Conn, error) {
-		return NewK8SExecConn(ctx, pod, time.Second*ConnectTimeout)
-		// return NewK8SPortForwardConn(ctx, s, time.Second*30)
-	})
+func (r *Runtime) UpdateDeployInfo() error {
+	info, err := r.loadRemoteDeployInfo()
+	if err != nil {
+		return fmt.Errorf("loading remote gadget info: %w", err)
+	}
+
+	return deployinfo.Store(info)
+}
+
+func (r *Runtime) loadRemoteDeployInfo() (*deployinfo.DeployInfo, error) {
+	ctx, cancelDial := context.WithTimeout(context.Background(), time.Second*ConnectTimeout)
+	defer cancelDial()
 
-	conn, err := grpc.DialContext(ctx, "", dialOpt, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock())
+	// use default params for now
+	params := r.ParamDescs().ToParams()
+	client, err := r.getClientFromRandomTarget(ctx, params)
 	if err != nil {
-		return nil, fmt.Errorf("dialing gadget pod on node %q: %w", pod.node, err)
+		return nil, fmt.Errorf("dialing random target: %w", err)
 	}
-	client := api.NewGadgetManagerClient(conn)
-	defer conn.Close()
```

Comment:
When is `conn` closed? I think `getClientFromRandomTarget` should return the connection instead of the client, so the caller has the opportunity to call `defer conn.Close()`. (opened https://github.com/inspektor-gadget/inspektor-gadget/pull/2092 to fix it)